### ISSUE TITLE
Too much overhead for training language model

### ISSUE THREAD

**Issue by qiuwei** [allenai/allennlp#2084]
Please first search our GitHub repository for similar questions.  If you don't find a similar example you can use the following template:

**System (please complete the following information):**
 - OS: Linux
 - Python version: 3.6.1
 - AllenNLP version: I installed from master"
 - PyTorch version: 0.4

**Question**

* I am training a character level language model(https://github.com/zalandoresearch/flair) using allennlp. I am using character tokenizer to prepare the dataset, following the language_modeling.py in the repository. However, creating one token for each char seems to have too much overhead, since there can be easily billions of chars for training a language model. 
For example, now I am training a model on a rather small dataset(which consists of 570537619 characters), the tokenization step already took hours.  Am I doing it wrong? Is there any way to get around this problem?

**Comment by sai-prasanna** [comment]
@qiuwei  Have you tried using laziness instead of trying to process the entire dataset?

Refer  https://github.com/allenai/allennlp/blob/master/tutorials/how_to/laziness.md

**Comment by matthew-z** [comment]
Yes, laziness will only increase the whole training time (it results in tokenization for every epoch), but it is a workaround that let the training start immediately.

**Comment by schmmd** [comment]
@brendan-ai2 now that we've ported the ELMo training to AllenNLP--is this still an issue?

Documentation: https://github.com/allenai/allennlp/blob/master/tutorials/how_to/training_transformer_elmo.md

**Comment by brendan-ai2** [comment]
@qiuwei, @schmmd, while training the Transformer ELMo reimplementation I used the `elmo_characters` token indexer. It's definitely plenty efficient. When I used the `characters` indexer (and thus the `CharacterTokenizer`) I did hit issues with memory usage during training. My hypothesis was that this was due to the tokenizer not truncating long words, but this is unverified. I'd suggest using `elmo_characters` if you need a quick fix.

@qiuwei, by any chance do you have a minimal repro of this issue? Especially if the issue is strictly during tokenization I would be interested in understanding the problem deeper.

Transformer ELMo config for reference: https://github.com/allenai/allennlp/blob/master/training_config/bidirectional_lm.jsonnet#L21

**Comment by qiuwei** [comment]
@brendan-ai2 I will try elmo_characters  out and keep you posted.
